vernac/docstring.ml

```diff
@@ -0,0 +1,117 @@
+(************************************************************************)
+
+(*i*)
+open Names
+open Globnames
+open Libobject
+(*i*)
+
+type docstring_kind =
+  | ModPathDocstring of ModPath.t
+  | GlobRefDocstring of GlobRef.t
+
+let subst_docstring_kind subst = function
+  | ModPathDocstring mp -> ModPathDocstring (Mod_subst.subst_mp subst mp)
+  | GlobRefDocstring gr -> GlobRefDocstring (fst (subst_global subst gr))
+
+let _eq_docstring_kind k1 k2 =
+  match k1, k2 with
+  | ModPathDocstring mp1, ModPathDocstring mp2 -> ModPath.equal mp1 mp2
+  | GlobRefDocstring gr1, GlobRefDocstring gr2 -> Environ.QGlobRef.equal (Global.env ()) gr1 gr2
+  | (ModPathDocstring _ | GlobRefDocstring _), _ -> false
+
+let string_of_docstring_kind = function
+  | ModPathDocstring mp -> DirPath.to_string (Nametab.dirpath_of_module mp)
+  | GlobRefDocstring gr -> Libnames.string_of_path (Nametab.path_of_global gr)
+
+module DocstringKind =
+  struct
+    type t = docstring_kind
+    let compare = Stdlib.compare
+  end
+
+type docstring_info = {
+  comment : string;
+  warning : bool;
+}
+
+module DocstringMap = Map.Make(DocstringKind)
+
+let docstring_table =
+  Summary.ref (DocstringMap.empty : docstring_info list DocstringMap.t) ~name:"docstring-table"
+
+let cache_docstring (_, k, m) =
+  let l = try DocstringMap.find k !docstring_table with Not_found -> [] in
+  docstring_table := DocstringMap.add k (m::l) !docstring_table
+
+let load_docstring _ x =
+  cache_docstring x
+
+let open_docstring _ _ = ()
+
+let subst_docstring (subst, (local, k, m)) =
+  (local, subst_docstring_kind subst k, m)
+
+let discharge_docstring (local, k, m as x) =
+  if local then None else Some x
+
+let classify_docstring (local, _, _) =
+  if local then Dispose else Substitute
+
+type docstring = bool * docstring_kind * docstring_info
+
+let inDocstring : docstring -> obj =
+  declare_object
+    { (default_object "DOC-STRINGS") with
+      cache_function = cache_docstring;
+      load_function = load_docstring;
+      open_function = simple_open ~cat:Hints.hint_cat open_docstring;
+      classify_function = classify_docstring;
+      discharge_function = discharge_docstring;
+      subst_function = subst_docstring }
+
+let declare_docstring_modpath mp info =
+  if info.warning then
+  Lib.add_leaf (inDocstring (false,ModPathDocstring mp,info))
+
+let declare_docstring gr info =
+  let local = isVarRef gr && Lib.is_in_section gr in
+  Lib.add_leaf (inDocstring (local,GlobRefDocstring gr,info))
+
+let warn_docstring =
+  let warnings = ref CString.Map.empty in
+  let pp s = Pp.str s in
+  fun ?loc (k,s) ->
+    let name = string_of_docstring_kind k in
+    let w =
+      match CString.Map.find_opt name !warnings with
+      | Some w -> w
+      | None ->
+        let info_cat = CWarnings.create_category ~name:"info" () in
+        let main_name = "info-warning" in
+        let main_cat, main_w = CWarnings.(create_hybrid ~name:main_name ~from:[info_cat] ()) in
```

Comment:
With your change, it gives the warnings `info-MyRoot.MyFile,info,default` which seems ok.

I'm not familiar with the design of the warning and deprecation system. What should it be for deprecation? Should it be `deprecated-libraries-since-8.XX,deprecated-since-8.XX,deprecated-libraries,deprecated,default` (on the model of what is done for abbreviations)? Something different?

Reply:
I pushed a new version that includes deprecation. As of now, it works as follows:
- in file foo.v:
  ```
  Library Warning "This implements foo arithmetic".
  Library Deprecation since "8.18" "Please use bar arithmetic".
  ```
- output at require time:
  ```
  Toplevel input, characters 0-12:
  > Require Foo.
  > ^^^^^^^^^^^^
  Warning: Library Foo is deprecated since 8.4.
  Please use bar arithmetic
  [deprecated-libraries-since-8.4,deprecated-since-8.4,deprecated-libraries,deprecated,default]
  Toplevel input, characters 0-12:
  > Require Foo.
  > ^^^^^^^^^^^^^
  Warning: This implements foo arithmetic [info-Foo,info,default]
  ```

PS: To be clear, I'm not attached to any syntax, nor to any particular solution. My motivation is primarily that this is needed for other PRs.